src/mds/Server.cc

```diff
@@ -2197,6 +2197,11 @@ void Server::early_reply(MDRequestRef& mdr, CInode *tracei, CDentry *tracedn)
     return;
   }
 
+  if (!mdr->prealloc_inos.empty()) {
+    dout(10) << "early_reply - prealloc inos, not allowed" << dendl;
+    return;
+  }
+
```

Comment:
IMO this makes no sense. Because the `mdr->prealloc_inos` hasn't been applied to the corresponding session yet, so the session is not dirty and won't delegate the inode in `mdr` to clients ?

So it should be safe to delegate inos in the session to clients.

Reply:
Since we are revisiting this change I would double check, but from the tests that I did when implementing this (IIRC), the preallocated inode list was being sent to the client in early reply and since this change ensured that the preallocated list would be persisted before the clients can use it, this check ensured that semantics.

But, as I said, I will double check. Thanks @lxbsz

Reply:
Sure. This has been long time ago and I will check it again.